### PR TITLE
[pigeon] Fixed decoding code for null-safe classes.

### PR DESCRIPTION
**Pull request by gaaclarke**
issue: https://github.com/flutter/flutter/issues/76405

## Pre-launch Checklist

- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated `CHANGELOG.md` to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the `#hackers-new` channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

### CODE REVIEW

File: packages/pigeon/lib/dart_generator.dart [hunk 1]
```diff
@@ -178,11 +178,11 @@ void _writeFlutterApi(
...
   switch (dataType) {
     case 'List':
-      return 'List<Object$nullTag>';
+      return 'List<Object$nullTag>$nullTag';
     case 'Map':
-      return 'Map<Object$nullTag, Object$nullTag>';
+      return 'Map<Object$nullTag, Object$nullTag>$nullTag';
     default:
-      return dataType;
+      return '$dataType$nullTag';
```
Comment (gaaclarke) [inline packages/pigeon/lib/dart_generator.dart:185]: This is the actual fix, other stuff is just testing infrastructure / tests.

File: packages/pigeon/platform_tests/flutter_null_safe_unit_tests/lib/null_safe_pigeon.dart [hunk 2]
```diff
@@ -0,0 +1,166 @@
+// Autogenerated from Pigeon (v0.1.21), do not edit directly.
+// See also: https://pub.dev/packages/pigeon
+// ignore_for_file: public_member_api_docs, non_constant_identifier_names, avoid_as, unused_import
+// @dart = 2.12
```
Comment (gaaclarke) [inline packages/pigeon/platform_tests/flutter_null_safe_unit_tests/lib/null_safe_pigeon.dart:4]: I had to add this file otherwise CI analysis steps fail.

**Comment by ditman** [comment]
Let me add @blasten to this review, he pointed yesterday to the code you're touching here!

File: packages/pigeon/platform_tests/flutter_null_safe_unit_tests/lib/null_safe_pigeon.dart (hunk 2 above)
Comment (blasten) [reply packages/pigeon/platform_tests/flutter_null_safe_unit_tests/lib/null_safe_pigeon.dart:4]: Maybe, we disable the analysis for `flutter_null_safe_unit_tests`? https://dart.dev/guides/language/analysis-options#the-analysis-options-file

File: packages/pigeon/lib/dart_generator.dart (hunk 1 above)
Comment (blasten) [reply packages/pigeon/lib/dart_generator.dart:185]: makes sense

**Review by blasten — APPROVED** [review]

File: packages/pigeon/platform_tests/flutter_null_safe_unit_tests/lib/null_safe_pigeon.dart (hunk 2 above)
Comment (gaaclarke) [reply packages/pigeon/platform_tests/flutter_null_safe_unit_tests/lib/null_safe_pigeon.dart:4]: Thanks for the pointer.  I'm on the fence as to whether it is a good idea or not.  I think I'll keep it this way for now.

**Pull request merged**